feat: validate new certificate on connect

didier-wenzek opened the pull request:
## Proposed changes

Let thin-edge maintain 2 certificates (*current* and *new*) for each cloud profile.
The paths to these certificates are derived from the `device.cert_path` tedge config setting.
  - `"$(tedge config get device.cert_path)"` is the path to the certificate currently used to connect the cloud endpoint
  - `"$(tedge config get device.cert_path).new"` is the path to a new certificate still to be validated (if any).

The command `tedge cert renew` does no more replace the current certificate but store the new certificate in its own file for further validation. The promotion of a new certificate as the current one is done by the `tedge connect` command after a successful connection.

- [x]  `tedge cert renew` doesn't replace the current certificate but creates a new file.
- [x]  `tedge connect c8y` validates any new certificate before promoting it as the current certificate
- [ ]  ~`tedge connect az` validates any new certificate before promoting it as the current certificate~
- [ ]  ~`tedge connect aws` validates any new certificate before promoting it as the current certificate~
- [x] 'tedge cert show --new' displays the new certificate instead of the current one
- [x] `tedge connect --offline` ignores with warning any new certificate
- [x] `tedge cert upload` uploads the new certificate when there is one
- [x] `tedge cert remove` removes removes new and current certificates
- [ ] ~`tedge cert needs-renewal --new` checks the new certificate if any~. useless
- [ ] ~Should `tedge cert upload` uploads the current *and* the new certificate when there are both?~ no, useless
- [x] On success, `tedge cert renew` shows the new certificate
- [x] On success, `tedge cert renew` warns that the device must be reconnected
- [x] `tedge cert renew` uses the CA given by the `--ca` option (`c8y` by default)


## Types of changes

<!--
What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_
-->

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue

https://github.com/thin-edge/thin-edge.io/issues/3541

## Checklist

<!--
_Put an `x` in the boxes that apply. You can also fill these out after
creating the PR. If you're unsure about any of them, don't hesitate to ask.
We're here to help! This is simply a reminder of what we are going to look for
before merging your code._
-->

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s. You can activate automatic signing by running `just prepare-dev` once)
- [x] I ran `just format` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `just check` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

reubenmiller commented:
@didier-wenzek I just tried out some of the commands (what is implemented so far), and it works nicely...it will definitely simplify the integration with custom renewers.

We can discuss the open points shortly, to finalize the direction of the PR.

didier-wenzek commented:
>  [ ] Who/What is responsible for deleting the .new certificate if the connection fails?

I think the better is to not delete the .new certificate when the connection is not successful.
- the root cause for the connection failure might be unrelated to the .new certificate
- the renew command will in any case replace the  .new certificate

>  [ ] What should the exit code be when using `tedge reconnect c8y` but switching to the .new certificate fails (for whatever reason). Ideally it should be a non-zero exit code. See the shell snippets for how having a non-zero exit code would help...

Offline, we agreed to exit with a specific status code when the connection failed with the .new certificate.

- 0: successfully connected (using the new certificate, if one was present)
- 1: failed to connect (possibly with the new certificate despite, this certificate has been successfully  validated)
- 2: command line parsing error (clap)
- 3: properly connected - but with the previous certificate while a new one is present

didier-wenzek commented:
> > [ ] What should the exit code be when using `tedge reconnect c8y` but switching to the .new certificate fails (for whatever reason). Ideally it should be a non-zero exit code. See the shell snippets for how having a non-zero exit code would help...
> 
> Offline, we agreed to exit with a specific status code when the connection failed with the .new certificate.
> 
>     * 0: successfully connected (using the new certificate, if one was present)
> 
>     * 1: failed to connect (possibly with the new certificate despite, this certificate has been successfully  validated)
> 
>     * 2: command line parsing error (clap)
> 
>     * 3: properly connected - but with the previous certificate while a new one is present

Done: https://github.com/thin-edge/thin-edge.io/pull/3551/commits/8be41908311c66b6532c6d3adaf585a6974a421a

reubenmiller commented:
@didier-wenzek Overall it works nicely (especially the new exit code for the `tedge reconnect c8y`).

There was only 1 item which could be improved:

### Item 1: tedge cert upload uploads the new certificate when there is one

If there is a .new certificate and the `tedge cert upload c8y` uploads the .new certificate instead of the default one, then the command should include some information about which certificate it uploaded, otherwise it is not super clear for the user, not even with `--log-level trace`.

Below shows the current output when there was both a "main" certificate and a .new certificate.

```
tedge cert upload c8y
Certificate uploaded successfully.
```

rina23q commented:
`tedge connect c8y` error message looks repeating. 
```
Checking new certificate validity: /etc/tedge/device-certs/tedge-certificate.pem.new... ✗
error: Connection error while creating device in Cumulocity: Mqtt state: Mqtt serialization/deserialization error: IO: received fatal alert: CertificateUnknown: Mqtt serialization/deserialization error: IO: received fatal alert: CertificateUnknown: IO: received fatal alert: CertificateUnknown: received fatal alert: CertificateUnknown
Error validating the new certificate: Connection error while creating device in Cumulocity: Mqtt state: Mqtt serialization/deserialization error: IO: received fatal alert: CertificateUnknown: Mqtt serialization/deserialization error: IO: received fatal alert: CertificateUnknown: IO: received fatal alert: CertificateUnknown: received fatal alert: CertificateUnknown
  => keep using the current certificate unchanged /etc/tedge/device-certs/tedge-certificate.pem
Creating device in Cumulocity cloud... ✗
error: Connection error while creating device in Cumulocity: Mqtt state: Mqtt serialization/deserialization error: IO: received fatal alert: CertificateUnknown: Mqtt serialization/deserialization error: IO: received fatal alert: CertificateUnknown: IO: received fatal alert: CertificateUnknown: received fatal alert: CertificateUnknown
```

The steps I have done is below. I didn't upload both old and new self-signed certificate with intention to cause an error.
```shell
root@11d89e85c51e:/setup# tedge cert remove
Certificate was successfully removed

root@11d89e85c51e:/setup# tedge cert create --device-id aaa
Certificate was successfully created

Certificate:   /etc/tedge/device-certs/tedge-certificate.pem
Subject:       CN=aaa, O=Thin Edge, OU=Device
Issuer:        CN=aaa, O=Thin Edge, OU=Device
Status:        VALID (expires in: 11months 29days 3h 50m 23s)
Valid from:    Thu, 10 Apr 2025 13:04:23 +0000
Valid until:   Fri, 10 Apr 2026 13:04:23 +0000
Serial number: 525188842049310620321124847034595044436808047873 (0x5bfe483c5c128f4bc6d974814f1e3f42ec2c1d01)
Thumbprint:    B2B6EBA9118F82A3FF188F5960D7EB8D46F36B3A

root@11d89e85c51e:/setup# tedge cert renew --self-signed
Certificate was successfully renewed, for un-interrupted service, the certificate has to be uploaded to the cloud

# I didn't upload the new certificate
root@11d89e85c51e:/setup# tedge connect c8y
connect to Cumulocity cloud.:
        device id: aaa
        cloud profile: <none>
        cloud host: rina.xxx.xxx:8883.  #masked
        auth type: Certificate
        certificate file: /etc/tedge/device-certs/tedge-certificate.pem
        cryptoki: false
        bridge: mosquitto
        service manager: systemd
        mosquitto version: 2.0.11
Checking new certificate validity: /etc/tedge/device-certs/tedge-certificate.pem.new... ✗
error: Connection error while creating device in Cumulocity: Mqtt state: Mqtt serialization/deserialization error: IO: received fatal alert: CertificateUnknown: Mqtt serialization/deserialization error: IO: received fatal alert: CertificateUnknown: IO: received fatal alert: CertificateUnknown: received fatal alert: CertificateUnknown
Error validating the new certificate: Connection error while creating device in Cumulocity: Mqtt state: Mqtt serialization/deserialization error: IO: received fatal alert: CertificateUnknown: Mqtt serialization/deserialization error: IO: received fatal alert: CertificateUnknown: IO: received fatal alert: CertificateUnknown: received fatal alert: CertificateUnknown
  => keep using the current certificate unchanged /etc/tedge/device-certs/tedge-certificate.pem
Creating device in Cumulocity cloud... ✗
error: Connection error while creating device in Cumulocity: Mqtt state: Mqtt serialization/deserialization error: IO: received fatal alert: CertificateUnknown: Mqtt serialization/deserialization error: IO: received fatal alert: CertificateUnknown: IO: received fatal alert: CertificateUnknown: received fatal alert: CertificateUnknown
```

reubenmiller commented:
> `tedge connect c8y` error message looks repeating.
> 

This isn't actually repeating, there is one error per certificate (in this case 2 attempts where both failed).

didier-wenzek commented:
> > error: Connection error while creating device in Cumulocity: Mqtt state: Mqtt serialization/deserialization error: IO: received fatal alert: CertificateUnknown: Mqtt serialization/deserialization error: IO: received fatal alert: CertificateUnknown: IO: received fatal alert: CertificateUnknown: received fatal alert: CertificateUnknown
> 
> Sorry what I meant "repeating" is a different point... one error message contains 4 times `received fatal alert: CertificateUnknown`. Is it actually one error that has many nests?
> 
> ```
> error: Connection error while creating device in Cumulocity: Mqtt state: Mqtt serialization/deserialization error: IO: received fatal alert: CertificateUnknown: Mqtt serialization/deserialization error: IO: received fatal alert: CertificateUnknown: IO: received fatal alert: CertificateUnknown: received fatal alert: CertificateUnknown
> ```

Truly fatal ;-) I will see if I can fix that (this is not a new error message).

didier-wenzek commented:
> If there is a .new certificate and the `tedge cert upload c8y` uploads the .new certificate instead of the default one, then the command should include some information about which certificate it uploaded, otherwise it is not super clear for the user, not even with `--log-level trace`.


Improved: https://github.com/thin-edge/thin-edge.io/pull/3551/commits/64c13dc61f0c2de3b9e3bc7e5a42187189bc9528

rina23q commented on crates/core/tedge/src/cli/connect/command.rs:
```diff
@@ -219,13 +212,88 @@ impl ConnectCommand {
+            }
+
+            create_device_with_direct_connection(
+                use_basic_auth,
```
This `use_basic_auth` seems always false. You can use `bridge_config.auth_type` to check if it is Basic or not.

didier-wenzek replied:
`use_basic_auth` is set `false` on purpose here, because if we reach this point this is because there are two certificates: a current one + a new one to be validated.

rina23q reviewed (commented):
I have code suggestions/questions.

reubenmiller commented on crates/core/tedge/src/cli/certificate/shift.rs:
```diff
@@ -0,0 +1,56 @@
+    pub fn new_certificate_path(cert_path: &Utf8Path) -> Utf8PathBuf {
+        let mut new_cert_path = cert_path.to_owned();
+        new_cert_path.set_file_name(match cert_path.file_name() {
+            None => ".new".to_string(),
```
Talking about the functional level, the `.new` file doesn't exist on it's own right? It is only suffix added to the cert path. If the cert path is empty then you might end up with a file just called `.new` but this case does not happen in practice as the cert path is always defined.

didier-wenzek replied:
Applied @rina23q 's suggestion to name this file `certificate.new` just to avoid to get a hidden file if our assumption that this should not happen is false.

https://github.com/thin-edge/thin-edge.io/pull/3551/commits/c2aa447a1a44102803ae79087671f46ba84c5caf

didier-wenzek commented on crates/core/tedge/src/cli/certificate/remove.rs:
```diff
@@ -33,6 +33,8 @@ impl Command for RemoveCertCmd {
 impl RemoveCertCmd {
     pub(crate) async fn remove_certificate(&self) -> Result<RemoveCertResult, CertError> {
+        let _new_cert_silently_removed =
+            fs::remove_file(&CertificateShift::new_certificate_path(&self.key_path)).await;
```
Oops https://github.com/thin-edge/thin-edge.io/pull/3551/commits/8c286033a6364c86da0e18f13a6eca43c0502c19

reubenmiller commented:
We'll look at merging this next week. I just spotted a minor UX thing when running the reconnect command locally, where the attempt/retry progress is mixed with the progress indicator line:

```
Disabling tedge-mapper-c8y... ✓
reconnect to Cumulocity cloud.:
        device id: TST_link_yellow_pistol
        cloud profile: <none>
        cloud host: thin-edge-io.eu-latest.cumulocity.com:8883
        auth type: Certificate
        certificate file: /etc/tedge/device-certs/tedge-certificate.pem
        cryptoki: false
        bridge: mosquitto
        service manager: systemd
        mosquitto version: 2.0.11
Connection test failed, attempt 1 of 3c/tedge/device-certs/tedge-certificate.pem.new... ⠹
Connection test failed, attempt 2 of 3c/tedge/device-certs/tedge-certificate.pem.new... ⠼
Checking new certificate validity: /etc/tedge/device-certs/tedge-certificate.pem.new... ✗
```

didier-wenzek commented:
> We'll look at merging this next week. I just spotted a minor UX thing when running the reconnect command locally, where the attempt/retry progress is mixed with the progress indicator line:

Fixed: https://github.com/thin-edge/thin-edge.io/pull/3551/commits/968c67b614853279085ec69c1567ae5a7e26185f

albinsuresh reviewed (commented):
LGTM. From the UX side of things, I found the following points unintuitive or less natural:

1. `tedge cert renew` command without the `c8y` argument implicitly doing the renewal with C8Y CA. Although it was done to seamlessly transition existing self-signed certs to CA signed, it feels like it'd surprise a few. I was wondering if it'd be better to keep the old behaviour, with an additional `WARNING` emitted on the console, informing the users of the CA signed option that's available and encouraging them to use that option instead.
2. `tedge cert renew` command is not giving any hints to the user that he needs to reconnect to the cloud, for the updated cert to take effect. Isn't that required?
3. A user running `tedge cert show` immediately after `tedge cert renew` might be surprised to see the old cert details instead of the new one, esp someone who's used to the old behaviour. They would be clueless until they discover the `--new` option. It'd be better if the `renew` command output informs the user of the presence of 2 certificates and encourage the user to use `--new` option with `tedge cert show`.

didier-wenzek commented on crates/core/tedge/src/cli/connect/command.rs:
```diff
@@ -219,13 +212,113 @@ impl ConnectCommand {
+        }
+
+        println!(
+            "The new certificate is now the current certificate {}",
```
fixed: https://github.com/thin-edge/thin-edge.io/pull/3551/commits/02ac42a8a3f8bf43f583cffa4e074307904986c5

didier-wenzek commented on crates/core/tedge/src/bridge/c8y.rs:
```diff
@@ -191,7 +191,7 @@ impl From<BridgeConfigC8yParams> for BridgeConfig {
             topics,
             bridge_location,
-            connection_check_attempts: 1,
+            connection_check_attempts: 3,
```
fixed: https://github.com/thin-edge/thin-edge.io/pull/3551/commits/3e0d2b6244e0b80bdd61bf53d68a0f08f0ca2aaa

reubenmiller commented:
One more slight unexpected behavior was found that after renewing the certificate (via the c8y certificate-authority), calling `tedge connect c8y` whilst already connected would result in swapping existing cert with the new one. I'm not 100% sure if the full connectivity check was done or not in the scenario (at least the console output does not show it).

```
# tedge cert renew c8y
root@45535a9e3668:/setup# tedge connect c8y
connect to Cumulocity cloud.:
        device id: TST_separate_equidistant_rink
        cloud profile: <none>
        cloud host: iot.latest.stage.c8y.io:8883
        auth type: Certificate
        certificate file: /etc/tedge/device-certs/tedge-certificate.pem
        cryptoki: false
        bridge: mosquitto
        service manager: systemd
        mosquitto version: 2.0.11
Validating new certificate: /etc/tedge/device-certs/tedge-certificate.pem.new... ✓
The new certificate is now the current certificate /etc/tedge/device-certs/tedge-certificate.pem
error: Connection is already established. To remove the existing connection, run `tedge disconnect c8y` and try again.
root@45535a9e3668:/setup# tedge connect c8y
```

didier-wenzek commented:
>     1. `tedge cert renew` command without the `c8y` argument implicitly doing the renewal with C8Y CA. Although it was done to seamlessly transition existing self-signed certs to CA signed, it feels like it'd surprise a few. That behaviour makes sense when someone is using the new `renew c8y` command but not when they run the existing `renew` command. I was wondering if it'd be better to keep the old behaviour, with an additional `WARNING` emitted on the console, informing the users of the CA signed option that's available and encouraging them to use that option instead.

The `tedge cert renew` command is related to two different end-points: the tenant of which the certificate has to renewed and the CA to which the renewal request has to be sent. As the most common case is to renew a certificate for a tenant using the CA of that tenant, the latter is by default derived from the former.

- `tedge cert renew c8y` renews the certificate used  to connect the c8y cloud profile using the c8y CA
- `tedge cert renew c8y --self-signed` renews the certificate to connect the c8y cloud profile using a self-signed certificate
- `tedge cert renew` renews the shared certificate using the c8y CA (the only one supported for now)
- `tedge cert renew c8y --self-signed` renews the shared certificate using a self-signed certificate

One option to make things more explicit could be to add an `--c8y-ca` option similar to the current `--self-signed` option.

- `tedge cert renew c8y --c8y-ca`              <= The CA can here be inferred from the cloud profile
- `tedge cert renew c8y --self-signed`
- `tedge cert renew --c8y-ca`                     <= this would notably makes sense for this case.
- `tedge cert renew c8y --self-signed`


Later, we could then introduce a `--aws-ca` to renew the certificate from AWS. Leading to more specific use cases:

- `tedge cert renew c8y --aws-ca` to renew the certificate used  to connect c8y using AWS CA

reubenmiller commented:
> 2. `tedge cert renew` command is not giving any hints to the user that he needs to reconnect to the cloud, for the updated cert to take effect. Isn't that required?
> 3. A user running `tedge cert show` immediately after `tedge cert renew` might be surprised to see the old cert details instead of the new one, esp someone who's used to the old behaviour. They would be clueless until they discover the `--new` option. It'd be better if the `renew` command output informs the user of the presence of 2 certificates and encourage the user to use `--new` option with `tedge cert show`.

Though this is mostly going to be an automated process, should a user manually renew the certificate, then it could be beneficial to provide some user instructions considering the problem that would arise if they forget to run `tedge reconnect` (e.g. manually renewing in an emergency situation where the automated renew service is not working for unexpected reasons, and it has to be renewed before it expires).

Below is a proposal, yes it is very verbose, but the output is targeted for users not an automated service. Automated services can always discard the output if it wants to reduce noise.

```
$ tedge cert renew c8y
```

*Output*

```sh
Certificate was successfully renewed, however you need to run 'tedge reconnect' to start using the new certificate.

The new certificate will only replace the current certificate if the validation is successful on the next tedge reconnect.

New certificate candidate information (also available when using using 'tedge cert show --new'):

Certificate:   /etc/tedge/device-certs/tedge-certificate.pem.new
Subject:       CN=TST_resonate_universal_trout, O=Thin Edge, OU=Device
Issuer:        C=United States, O=Cumulocity, CN=t9679
Status:        VALID (expires in: 11months 30days 3h 42m 27s)
Valid from:    Mon, 14 Apr 2025 15:19:52 +0000
Valid until:   Tue, 14 Apr 2026 15:19:52 +0000
Serial number: 1744643992955 (0x19634e2ad7b)
Thumbprint:    7F88BF6CCE14C9ACCF805F83298BB36FC39C8752
```

didier-wenzek commented:
> One more slight unexpected behavior was found that after renewing the certificate (via the c8y certificate-authority), calling `tedge connect c8y` whilst already connected would result in swapping existing cert with the new one. I'm not 100% sure if the full connectivity check was done or not in the scenario (at least the console output does not show it).

The full connectivity check was actually and properly done - but not the reconnect.  Indeed, confusing.

Fixed: https://github.com/thin-edge/thin-edge.io/pull/3551/commits/ff66294d8594afb106ddf2f6588ef3011dbb6947

reubenmiller commented:
> * tedge cert renew c8y --self-signed

I personally find having too many switches a bit more difficult to track (though this is a personal preference). What about setting a specific provider (thing that is actually renewing or regenerating the certificate).

```sh
tedge cert renew --provider <type>

tedge cert renew --provider self-signed
tedge cert renew c8y --provider self-signed
tedge cert renew c8y --provider c8y     <= default, when using 'tedge cert renew c8y'
tedge cert renew aws --provider aws     <= example of some kind of future provider
tedge cert renew c8y --provider aws     <= example of some kind of future provider
```

By having a flag that accepts a value is a bit more flexible in terms of UX to add/remove values rather than switch-like flags (e.g. --self-signed).

reubenmiller commented:
> > One more slight unexpected behavior was found that after renewing the certificate (via the c8y certificate-authority), calling `tedge connect c8y` whilst already connected would result in swapping existing cert with the new one. I'm not 100% sure if the full connectivity check was done or not in the scenario (at least the console output does not show it).
> 
> The full connectivity check was actually and properly done - but not the reconnect. Indeed, confusing.
> 
> Fixed: [ff66294](https://github.com/thin-edge/thin-edge.io/commit/ff66294d8594afb106ddf2f6588ef3011dbb6947)

Though running `tedge connect c8y` on an already connected device seems unexpected given that a direct connection attempt will most likely result in the existing MQTT connection being disconnected (depending on which core the MQTT connection is handled by)...this could be potentially disruptive/unexpected. 

For instance, if tedge is already connected to c8y, and you subscribe to the topics, then you might see the disconnection event which occurs due to two MQTT clients connecting to the cloud MQTT broker with the same client id. (below shows the output of the `tedge mqtt sub '#'` whilst running `tedge cert renew && tedge connect c8y`:

```
tedge mqtt sub '#'
[te/device/main/service/mosquitto-c8y-bridge/status/health] 0
[c8y/s/us] 102,TST_change_gray_denotation:device:main:service:mosquitto-c8y-bridge,service,mosquitto-c8y-bridge,down
[te/device/main/service/mosquitto-c8y-bridge/status/health] 1
[c8y/s/us] 102,TST_change_gray_denotation:device:main:service:mosquitto-c8y-bridge,service,mosquitto-c8y-bridge,up
```

And the current behaviour seems unexpected considering if you run `tedge connect c8y` multiple times, then it always results in the following error:

```sh
error: Connection is already established. To remove the existing connection, run `tedge disconnect c8y` and try again.
```

So should we change the behaviour to the following:

1. Check if already connected, then exit with an "Connection is already established" error 
2. If not already connected, then proceed with trying to connect (including detection of a new certifcate and testing that first).

albinsuresh commented:
> > * tedge cert renew c8y --self-signed
> 
> I personally find having too many switches a bit more difficult to track (though this is a personal preference). What about setting a specific provider (thing that is actually renewing or regenerating the certificate).
> 
> ```shell
> tedge cert renew --provider <type>
> 
> tedge cert renew --provider self-signed
> tedge cert renew c8y --provider self-signed
> tedge cert renew c8y --provider c8y     <= default, when using 'tedge cert renew c8y'
> tedge cert renew aws --provider aws     <= example of some kind of future provider
> tedge cert renew c8y --provider aws     <= example of some kind of future provider
> ```
> 
> By having a flag that accepts a value is a bit more flexible in terms of UX to add/remove values rather than switch-like flags (e.g. --self-signed).

Just to re-confirm the defaults:
* `tedge cert renew c8y` -> `tedge cert renew c8y --provider c8y`
* `tedge cert renew` -> `tedge cert renew --provider self-signed` OR `tedge cert renew --provider c8y` ?

Optional suggestion: how about naming that `--provider` option as `--ca` to make it more direct? Like `--ca c8y`,  `--ca self` etc so that the user won't have to lookup what a `provider` is?

didier-wenzek commented:
> So should we change the behaviour to the following:
> 1. Check if already connected, then exit with an "Connection is already established" error 
> 2. If not already connected, then proceed with trying to connect (including detection of a new certifcate and testing that first).

This is what has been implemented with https://github.com/thin-edge/thin-edge.io/commit/ff66294d8594afb106ddf2f6588ef3011dbb6947 . Am I missing something?

didier-wenzek commented:
> Just to re-confirm the defaults:
* `tedge cert renew c8y` -> `tedge cert renew c8y --provider c8y`  __yes__
* `tedge cert renew` -> `tedge cert renew --provider self-signed` OR __`tedge cert renew --provider c8y`__ ?

> Optional suggestion: how about naming that `--provider` option as `--ca` to make it more direct? Like `--ca c8y`,  `--ca self` etc so that the user won't have to lookup what a `provider` is?

Indeed, `--ca` sounds more appropriate. However I will not shorten `--ca self-signed`

reubenmiller commented:
> > So should we change the behaviour to the following:
> > 
> > 1. Check if already connected, then exit with an "Connection is already established" error
> > 2. If not already connected, then proceed with trying to connect (including detection of a new certifcate and testing that first).
> 
> This is what has been implemented with [ff66294](https://github.com/thin-edge/thin-edge.io/commit/ff66294d8594afb106ddf2f6588ef3011dbb6947) . Am I missing something?

No, it seems the current behaviour is to always do step 2 regardless of the outcome of step 1. E.g. if there is a new certificate it will validate it and replace the current certificate without checking if tedge is already connected or not.

didier-wenzek commented:
As agreed offline:

- [x] On success, `tedge cert renew` shows the new certificate
- [x] On success, `tedge cert renew` warns that the device must be reconnected
- [x] `tedge cert renew` uses the CA given by the `--ca` option (`c8y` by default)

I also updated `tedge cert create` and `tedge cert download` to have a consistent behavior.

Addressed by:
- https://github.com/thin-edge/thin-edge.io/pull/3551/commits/a2b4369644fbf83098818bc0583b1cf6b9554d51
- https://github.com/thin-edge/thin-edge.io/pull/3551/commits/bc3e6ece006f1acbe5deefa2a8b3360a226e5bea

reubenmiller reviewed (approved):
Approved. A very nice feature that makes the certificate renewal process a lot more reliable, and it makes integrating custom renew logic very easy.

albinsuresh reviewed (approved):
Everything is working nicely. The command outputs are really clear now.

**Not for this PR**, but just a thought: Should the `tedge cert show` command give any hints about the presence of the `.new` certificate if one is present (with an additional hint about how to view it using the `--new` option)? Since the `cert renew` command output is sufficiently verbose right now, this isn't such a strong requirement. Just an additional guidance, that's all.
